pkg/clients/cryptokey/cryptokey.go

```diff
@@ -0,0 +1,155 @@
+		}
+		if in.Primary.Attestation != nil {
+			o.Primary.Attestation = &v1alpha1.KeyOperationAttestation{
+				Content: in.Primary.Attestation.Content,
+				Format:  in.Primary.Attestation.Format,
+			}
+		}
+		if in.Primary.ExternalProtectionLevelOptions != nil {
+			o.Primary.ExternalProtectionLevelOptions = &v1alpha1.ExternalProtectionLevelOptions{
+				ExternalKeyUri: in.Primary.ExternalProtectionLevelOptions.ExternalKeyUri,
+			}
+		}
+	}
+
+	return o
+}
+
+// LateInitializeSpec fills unassigned fields with the values in cloudkms.CryptoKey object.
+func LateInitializeSpec(spec *v1alpha1.CryptoKeyParameters, in cloudkms.CryptoKey) {
+	spec.Labels = in.Labels
+	spec.RotationPeriod = gcp.LateInitializeString(spec.RotationPeriod, in.RotationPeriod)
+	spec.NextRotationTime = gcp.LateInitializeString(spec.NextRotationTime, in.NextRotationTime)
+	if in.VersionTemplate != nil {
+		if spec.VersionTemplate == nil {
+			spec.VersionTemplate = &v1alpha1.CryptoKeyVersionTemplate{}
+		}
+		spec.VersionTemplate.ProtectionLevel = gcp.LateInitializeString(
+			spec.VersionTemplate.ProtectionLevel, in.VersionTemplate.ProtectionLevel)
+		spec.VersionTemplate.Algorithm = gcp.LateInitializeString(
+			spec.VersionTemplate.Algorithm, in.VersionTemplate.Algorithm)
+	}
+}
+
+// IsUpToDate checks whether current state is up-to-date compared to the given
+// set of parameters.
+func IsUpToDate(name string, in *v1alpha1.CryptoKeyParameters, observed *cloudkms.CryptoKey) (bool, string, error) {
+	um := make([]string, 0, 10)
+	generated, err := copystructure.Copy(observed)
+	if err != nil {
+		return true, "", errors.Wrap(err, errCheckUpToDate)
+	}
+	desired, ok := generated.(*cloudkms.CryptoKey)
+	if !ok {
+		return true, "", errors.New(errCheckUpToDate)
+	}
+	GenerateCryptoKeyInstance(name, *in, desired)
+
+	if !cmp.Equal(desired.Labels, observed.Labels, cmpopts.EquateEmpty()) {
+		um = append(um, "labels")
+	}
+	if !cmp.Equal(desired.Purpose, observed.Purpose, cmpopts.EquateEmpty()) {
+		um = append(um, "purpose")
+	}
+	if !cmp.Equal(desired.RotationPeriod, observed.RotationPeriod, cmpopts.EquateEmpty()) {
+		um = append(um, "rotationPeriod")
+	}
+	if !cmp.Equal(desired.NextRotationTime, observed.NextRotationTime, cmpopts.EquateEmpty()) {
+		um = append(um, "nextRotationTime")
+	}
+
+	if !cmp.Equal(desired.VersionTemplate, observed.VersionTemplate, cmpopts.EquateEmpty(),
+		cmpopts.IgnoreFields(cloudkms.CryptoKeyVersionTemplate{}, "ForceSendFields"),
+		cmpopts.IgnoreFields(cloudkms.CryptoKeyVersionTemplate{}, "NullFields"),
+	) {
+		if !cmp.Equal(desired.VersionTemplate.Algorithm, observed.VersionTemplate.Algorithm, cmpopts.EquateEmpty()) {
+			um = append(um, "versionTemplate.algorithm")
+		}
+		if !cmp.Equal(desired.VersionTemplate.ProtectionLevel, observed.VersionTemplate.ProtectionLevel, cmpopts.EquateEmpty()) {
+			um = append(um, "versionTemplate.protectionLevel")
+		}
+
+	}
+
+	if len(um) > 0 {
+		return false, strings.Join(um, ","), nil
```

Comment:
Could be a potentially cleaner model to make `um` and actual data structure that is returned from this function. Maybe something like:

```
type Mask struct {
  labels bool
  purpose bool
  ...
}
```
Then set those booleans, return the struct, and have a `String()` method defined on it. Potentially a bit easier to test as well. However, in practice I think what you have here will work just fine so feel free to ignore suggestions :)

Reply:
Sounds more professional :) However, I would prefer to keep it simple and straightforward this time. Also, I didn't feel good about keeping another struct with one element per each config, at least I don't think it pays for the plumbing needed given what we need is simple. Thanks for the suggestion though.